src/java.base/macosx/classes/sun/nio/fs/MacOSXWatchService.java

```diff
@@ -0,0 +1,634 @@
+        private long         eventStreamRef;
+        private final Object eventStreamRefLock = new Object();
+
+        private DirectorySnapshot directorySnapshot;
+
+        MacOSXWatchKey(final MacOSXWatchService watchService,
+                       final UnixPath dir,
+                       final Object rootPathKey) throws IOException {
+            super(dir, watchService);
+            this.realRootPath       = dir.toRealPath().normalize();
+            this.realRootPathLength = realRootPath.toString().length() + 1;
+            this.rootPathKey        = rootPathKey;
+        }
+
+        synchronized void enable(final CFRunLoopThread runLoopThread,
+                                 final EnumSet<FSEventKind> eventsToWatch,
+                                 final EnumSet<WatchModifier> modifierSet) throws IOException {
+            assert(!isValid());
+
+            this.eventsToWatch = eventsToWatch;
+
+            directorySnapshot = DirectorySnapshot.create(getRealRootPath());
+
+            synchronized (eventStreamRefLock) {
+                final int kFSEventStreamCreateFlagWatchRoot  = 0x00000004;
+                eventStreamRef = MacOSXWatchService.eventStreamCreate(
+                        realRootPath.toString(),
+                        WatchModifier.sensitivityOf(modifierSet),
+                        kFSEventStreamCreateFlagWatchRoot);
+
+                if (eventStreamRef == 0)
+                    throw new IOException("Unable to create FSEventStream");
+
+                MacOSXWatchService.eventStreamSchedule(eventStreamRef,
+                        runLoopThread.getRunLoopRef());
+            }
+        }
+
+        synchronized void disable() {
+            invalidate();
+            directorySnapshot = null;
+        }
+
+        synchronized void handleEvents(final String[] paths, long eventFlagsPtr) {
+            if (paths == null) {
+                reportOverflow(null);
+                return;
+            }
+
+            if (updateNeeded(paths, eventFlagsPtr)) {
+                directorySnapshot.update(MacOSXWatchKey.this);
+            }
+        }
+
+        private Path toRelativePath(final String absPath) {
+            return   (absPath.length() > realRootPathLength)
+                    ? theFS.getPath(absPath.substring(realRootPathLength))
+                    : relativeRootPath;
+        }
+
+        private boolean updateNeeded(String[] paths, long eventFlagsPtr) {
+            boolean rootChanged = false;
+            for (final String absPath : paths) {
+                if (absPath == null) {
+                    reportOverflow(null);
+                    continue;
+                }
+
+                Path path = toRelativePath(absPath);
+
+                if (!relativeRootPath.equals(path)) {
+                    // Ignore events from subdirectories for now.
+                    continue;
+                }
+
+                final int flags = unsafe.getInt(eventFlagsPtr);
```

Comment:
This looks like the only use of `Unsafe`. Could the objective be accomplished without using it, e.g., changing `long eventFlagsPtr` to a `long[]` or, maybe even simpler, a `boolean[]`?

Reply:
Earlier on the mailing list @AlanBateman wrote "It would help the review if the native code is just simple wrappers about the FSEvents as we'll likely replace a lot of the JNI code in this area in time." 
I understood that as a suggestion to keep the native code as simple as possible and, conversely, do as much as possible on the Java side. Reading native memory here was one of the decisions that were based on that understanding. 
I may have, of course, read too much into it, so I'm not opposed to implementing your suggestion. I would like, however, to have a confirmation that this direction of the changes is OK with @AlanBateman also.

Reply:
Your understanding is correct, but sometimes precisely mapping the native API can result in more complication than needed. While this might not be a strictly analogous example, consider the BSD system call `setattrlist(2)`. We only need to use it at present to set file times, so this call was exposed at the Java level with a simpler set of parameters. See [sun.nio.fs.BsdNativeDispatcher::setattrlist](https://github.com/openjdk/jdk/blob/5285035ed9bb43a40108e4d046e0de317730f193/src/java.base/macosx/classes/sun/nio/fs/BsdNativeDispatcher.java#L84
). I’ll leave it to @AlanBateman to express his own opinion on this.